File: src/main/java/com/vc/medium/L250.scala
package com.vc.medium

class TreeNodeL250(var _value: Int) {
  var value: Int = _value
  var left: TreeNodeL250 = null
  var right: TreeNodeL250 = null
}

object L250 {
  def countUnivalSubtrees(root: TreeNodeL250): Int = {
    var count = 0
    def solve(root: TreeNodeL250, value: Int): Boolean = {
      if(root == null) return true
      if(!solve(root.left, root.value) | !solve(root.right, root.value)) return false
      count += 1
      root.value == value
    }
    solve(root, 0)
    count
  }
}

File: src/main/java/com/vc/medium/L250.java
package com.vc.medium;

import com.vc.hard.TreeNode;

class L250 {
    int res = 0;
    public int countUnivalSubtrees(TreeNode root) {
        helper(root);
        return res;
    }

    private boolean helper(TreeNode root) {
        if(root == null) return true;

        boolean left = helper(root.left);
        boolean right = helper(root.right);

        if(left && right
                && (root.right == null || root.right.val == root.val)
                && (root.left == null || root.left.val == root.val)) {
            res++;
            return true;
        }
        return false;
    }
}